atomix/cluster/src/test/java/io/atomix/raft/ControllableRaftContexts.java

```diff
@@ -310,35 +319,72 @@ public void clientAppendOnLeader() {
     }
   }
 
+  public void snapshotAndCompact(final MemberId memberId) {
+    final RaftContext raftContext = raftServers.get(memberId);
+    // Take snapshot at an index between lastSnapshotIndex and current commitIndex
+    final TestSnapshotStore testSnapshotStore = snapshotStores.get(memberId);
+    final var startIndex =
+        Math.max(raftContext.getLog().getFirstIndex(), testSnapshotStore.getCurrentSnapshotIndex());
+    if (startIndex >= raftContext.getCommitIndex()) {
```

Comment:
I didn't look into why it doesn't work. I didn't allow this case because I know that this is how Zeebe behaves. Zeebe conservatively always select previous index of the processedPosition. So it is guaranteed that atleast one committed entry after the snapshot index exist in the log. This is not a requirement for raft though. Raft should ideally allow taking and replicating snapshot at the commitIndex. May be let's keep it like this as it is how we use it.

Reply:
> I didn't allow this case because I know that this is how Zeebe behaves

> May be let's keep it like this as it is how we use it.

That makes sense, thanks :+1: